### PR TITLE
8332589: ubsan: unix/native/libjava/ProcessImpl_md.c:562:5: runtime error: null pointer passed as argument 2, which is declared to never be null

### PR DESCRIPTION
**Pull request by MBaesken**
When building with ubsan enabled (--enable-uban) on Linux x86_64 and doing jtreg tests afterwards I run into this error :

/jdk/src/java.base/unix/native/libjava/ProcessImpl_md.c:562:5: runtime error: null pointer passed as argument 2, which is declared to never be null
    #0 0x7fd95bec78d8 in spawnChild /jdk/src/java.base/unix/native/libjava/ProcessImpl_md.c:562
    #1 0x7fd95bec78d8 in startChild /jdk/src/java.base/unix/native/libjava/ProcessImpl_md.c:612
    #2 0x7fd95bec78d8 in Java_java_lang_ProcessImpl_forkAndExec /jdk/src/java.base/unix/native/libjava/ProcessImpl_md.c:712
    #3 0x7fd93797a06d (<unknown module>)

this is the memcpy call getting an unexpected null pointer :
    memcpy(buf+offset, c->pdir, sp.dirlen); gets a second parameter null.
Something similar was discussed and fixed here https://bugs.python.org/issue27570 for Python .

Similar issue in OpenJDK _ 
https://bugs.openjdk.org/browse/JDK-8332473
8332473: ubsan: growableArray.hpp:290:10: runtime error: null pointer passed as argument 1, which is declared to never be null

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8332589: ubsan: unix/native/libjava/ProcessImpl_md.c:562:5: runtime error: null pointer passed as argument 2, which is declared to never be null`

### Issue
 * [JDK-8332589](https://bugs.openjdk.org/browse/JDK-8332589): ubsan: unix/native/libjava/ProcessImpl_md.c:562:5: runtime error: null pointer passed as argument 2, which is declared to never be null (**Bug** - P4)

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19329/head:pull/19329` \
`$ git checkout pull/19329`

Update a local copy of the PR: \
`$ git checkout pull/19329` \
`$ git pull https://git.openjdk.org/jdk.git pull/19329/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19329`

View PR using the GUI difftool: \
`$ git pr show -t 19329`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19329.diff">https://git.openjdk.org/jdk/pull/19329.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19329#issuecomment-2122781210)

### CODE REVIEW

**Comment by magicus** [comment]
How come `c->pdir` can be null? Is it if `sp.dirlen` is 0?

**Comment by MBaesken** [comment]
> How come `c->pdir` can be null? Is it if `sp.dirlen` is 0?

I added a bit of tracing output and get 
**** Attention c->pdir is null; sp.dirlen is 0 
So yes sp.dirlen is also 0 .

File: src/java.base/unix/native/libjava/ProcessImpl_md.c [hunk 1]
```diff
@@ -559,7 +559,9 @@ spawnChild(JNIEnv *env, jobject process, ChildStuff *c, const char *helperpath)
     }
     offset = copystrings(buf, 0, &c->argv[0]);
     offset = copystrings(buf, offset, &c->envv[0]);
-    memcpy(buf+offset, c->pdir, sp.dirlen);
+    if (c->pdir != NULL) {
+        memcpy(buf+offset, c->pdir, sp.dirlen);
+    }
     offset += sp.dirlen;
```
Comment (RogerRiggs) [inline src/java.base/unix/native/libjava/ProcessImpl_md.c:565]: I'd be inclined to check sp.dirlen > 0 in the `if` and move the offset += inside too. Like:
```suggestion
    if (sp.dirlen > 0 && c->pdir != NULL) {
        memcpy(buf+offset, c->pdir, sp.dirlen);
        offset += sp.dirlen;
    }
```
The behavior is correct either way.

**Comment by MBaesken** [comment]
Hi Roger, thanks for the hint !  I adjusted the coding.

**Review by magicus — COMMENTED** [review]
This looks much better. 

However, if we ever call this with a non-zero `sp.dirlen` and a null `c->pdir`, we'd be in trouble. In the old code, we would have crashed. Now, we will just silently ignore this, and God knows what will happen after that part.

I don't have the full context on how this function is used, but if there is some way you can report back with an error to the caller in that case, I think it would be appropriate.

**Comment by MBaesken** [comment]
>However, if we ever call this with a non-zero sp.dirlen and a null c->pdir, we'd be in trouble. In the old code, we would have 
> crashed. Now, we will just silently ignore this, and God knows what will happen after that part.

In this special case we can return -1 ;  the spawnChild  function already returns -1 in some other error cases.

**Review by RogerRiggs — APPROVED** [review]
LGTM

**Comment by MBaesken** [comment]
Roger/Magnus,  should I still add some handling for Magnus' concern : "However, if we ever call this with a non-zero sp.dirlen and a null c->pdir, we'd be in trouble."
E.g. check for this and return -1  ?

**Comment by magicus** [comment]
As a general principle, I think you should detect errors and report them instead of ignoring them and continue. I don't know any details of this specific code, though.

The scenario with a null pointer and non-0 length seems unlikely and would only happen if there is a bug in the calling code. Nevertheless, I think it would be more prudent to guard against this condition.

**Comment by MBaesken** [comment]
Hi Magnus, I added handling of the special case that  memcpy src is a null pointer and non-0 length was given .

File: src/java.base/unix/native/libjava/ProcessImpl_md.c [hunk 2]
```diff
@@ -559,9 +559,16 @@ spawnChild(JNIEnv *env, jobject process, ChildStuff *c, const char *helperpath)
     }
     offset = copystrings(buf, 0, &c->argv[0]);
     offset = copystrings(buf, offset, &c->envv[0]);
-    if (sp.dirlen > 0 && c->pdir != NULL) {
-        memcpy(buf+offset, c->pdir, sp.dirlen);
-        offset += sp.dirlen;
+    if (c->pdir != NULL) {
+        if (sp.dirlen > 0) {
```
Comment (magicus) [inline src/java.base/unix/native/libjava/ProcessImpl_md.c:563]: As long as c->pdir is non-null, I think the code below is safe to execute. `memcpy(a, b, len)` should be okay if `len` is 0, as long as `a` and `b` are non-null, right?

So this check here is not needed, I think.

**Comment by magicus** [comment]
This looks much safer. Thank you! 

I think the code can be simplified a bit, as commented. It does not matter much, so you can keep the current code as well if you think it looks better.

**Review by RogerRiggs — APPROVED** [review]
LGTM  (Freeing the buffer on fail)

**Review by TheRealMDoerr — APPROVED** [review]
I also think that the `sp.dirlen > 0` check is not necessary, but I'm ok with this version, too.

**Pull request closed without merging**